src/Raven.Server/Documents/DocumentsStorage.cs

```diff
@@ -2563,5 +2618,380 @@ private CollectionName ExtractCollectionName(DocumentsOperationContext context,
+            result.ContentType = new LazyStringValue(null, ptr, size, context);
+
+            var streamIdentifier = tvr.Read((int) AttachmentsTable.StreamIdentifier, out size);
+            Slice.External(context.Allocator, streamIdentifier, size, out result.StreamIdentifier);
+
+            return result;
+        }
+
+        private static LazyStringValue TableValueToKey(JsonOperationContext context, int index, ref TableValueReader tvr)
+        {
+            int size;
+            // See format of the lazy string key in the GetLowerKeySliceAndStorageKey method
+            byte offset;
+            var ptr = tvr.Read(index, out size);
+            size = BlittableJsonReaderBase.ReadVariableSizeInt(ptr, 0, out offset);
+            return new LazyStringValue(null, ptr + offset, size, context);
+        }
+
+        private static void ThrowConcurrentExceptionOnMissingAttacment(string documentId, string name, long expectedEtag)
+        {
+            throw new ConcurrencyException(
+                $"Attachment {name} of '{documentId}' does not exists, but Put was called with etag {expectedEtag}. Optimistic concurrency violation, transaction will be aborted.")
+            {
+                ExpectedETag = expectedEtag
+            };
+        }
+
+        public void DeleteAttachment(DocumentsOperationContext context, string documentId, string name, long? expectedEtag)
+        {
+            if (string.IsNullOrWhiteSpace(documentId))
+                throw new ArgumentException("Argument is null or whitespace", nameof(documentId));
+            if (string.IsNullOrWhiteSpace(name))
+                throw new ArgumentException("Argument is null or whitespace", nameof(name));
+            if (context.Transaction == null)
+                throw new ArgumentException("Context must be set with a valid transaction before calling Get", nameof(context));
+
+            Slice lowerDocumentId, lowerName;
+            using (DocumentKeyWorker.GetSliceFromKey(context, documentId, out lowerDocumentId))
+            using (DocumentKeyWorker.GetSliceFromKey(context, name, out lowerName))
+            {
+                Slice keySlice;
+                using (GetAttachmentKey(context, lowerDocumentId.Content.Ptr, lowerDocumentId.Size, lowerName.Content.Ptr, lowerName.Size, out keySlice))
+                {
+                    DeleteAttachment(context, keySlice, lowerDocumentId, documentId, name, expectedEtag);
+                }
+            }
+        }
+
+        public void DeleteAttachment(DocumentsOperationContext context, Slice keySlice, Slice lowerDocumentId, string documentId, string name, long? expectedEtag)
+        {
+            Document document;
+            try
+            {
+                document = Get(context, lowerDocumentId);
+                if (document == null)
+                    throw new InvalidOperationException($"Cannot delete attachment {name} on a not exist document '{documentId}'.");
```

Comment:
'not exist' -> 'non existent'
And why is this an error? If they don't have an etag, that is fine and will be a noop